File: Data Structures & Algorithms/maximum-subarray/submission-1.java
class Solution {
    public int maxSubArray(int[] nums) {
        if (nums.length == 1) return nums[0];

        int [] sums = new int[nums.length];
        sums[0] = nums[0];
        int max = sums[0];
        for (int i = 1; i < nums.length; i++) {
            sums[i] = Math.max(sums[i - 1] + nums[i], nums[i]);
            max = Math.max(sums[i], max);
        }

        return max;
    }
}

File: Data Structures & Algorithms/house-robber/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    rob(nums) {
    if (!nums || nums.length == 0) return 0;
    if (nums.length == 1) return nums[0];
    const cache = new Int32Array(nums.length).fill(-1);
    cache[0] = nums[0];
    cache[1] = nums[1];
    let max = Math.max(cache[0], cache[1]);
    for (let i = 2; i < nums.length; i++) {

        let index = i - 2;
        let maxIndex = index;
        if (index - 1 >= 0) {
            maxIndex = cache[index] > cache[index - 1] ? index : index - 1;
        }

        cache[i] = nums[i] + cache[maxIndex];
        max = Math.max(max, cache[i]);
    }


    return max;
        
    }
}
